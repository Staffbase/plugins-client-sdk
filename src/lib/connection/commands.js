import reverse from '../utils/reverseObject';

/**
 * @typedef {{ ios: string, android: string, version: string, mobile: string, native: string, openLink: string}} PlatformCommands
 * @typedef {{ 'dev-ios': string, 'dev-android': string, 'app-version': string, 'app-mobile': string, 'app-native': string, 'openExternalLink':string}} ReversedPlatformCommands
 */

/**
 * Platform independat definition
 * of supported commands
 *
 * @type {PlatformCommands}
 */
export const commands = {
  ios: 'dev-ios',
  android: 'dev-android',
  version: 'app-version',
  mobile: 'app-mobile',
  native: 'app-native',
  openLink: 'openExternalLink',
  nativeUpload: 'nativeFileUpload'
};

/**
 * All supported commands with reversed keys and properties
 *
 * @type {ReversedPlatformCommands>}
 */
export const reversedCommands = reverse(commands);

export default commands;
