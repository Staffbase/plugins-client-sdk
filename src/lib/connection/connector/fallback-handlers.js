/**
 * Fallbacks for all sdk commands
 */
let log = require('loglevel');
let userAgent = navigator.userAgent || navigator.vendor || window.opera || '';
/**
 * Get the current Staffbase app version
 *
 * @return {String} version
 */
export function getVersion() {
  log.debug('fallback/getVersion');
  return '3.4';
}

/**
 * Are we running in a native app
 *
 * works only for ios
 * @return {Boolean}
 */
export function isNative() {
  log.debug('fallback/isNative');
  let safari = /safari/i.test(userAgent);
  return !safari && isIos();
}

/**
 * Are we running on a mobile device
 *
 * @return {Boolean}
 */
export function isMobile() {
  log.debug('fallback/isMobile');
  return /Android|webOS|iPhone|iPad|BlackBerry|BB10|IEMobile|Opera Mini/i.test(userAgent);
}

/**
 * Are we running on a desktop device
 *
 * @return {Boolean}
 */
export function isDesktop() {
  log.debug('fallback/isDesktop');
  return /Win|Mac|Linux/i.test(userAgent);
}

/**
 * Are we running on android
 *
 * @return {Boolean}
 */
export function isAndroid() {
  log.debug('fallback/isAndroid');
  return /Android/i.test(userAgent);
}

/**
 * Are we running on ios
 *
 * @return {Boolean}
 */
export function isIos() {
  log.debug('fallback/isIos');
  return /iPhone|iPad|iPod/i.test(userAgent);
}

/**
 * Open an external link
 *
 * @param {String} url address
 */
export function openLink(url) {
  log.debug('fallback/openLink');
  window.open(url, '_blank');
}

/**
 * Open an external link
 *
 * @param {String} url address
 */
export function nativeUpload(url) {
  log.debug('fallback/nativeUpload');
  log.warn('Native upload is not possible in fallback mode.');
  // nothing we can do here
  return;
}

/**
 * Get extensive locale information.
 *
 * @return {Object} containing various language informations
 */
export function langInfos() {
  log.debug('fallback/langInfos');
  return {
    contentLanguages: ['en_EN', 'de_DE']
  };
}
