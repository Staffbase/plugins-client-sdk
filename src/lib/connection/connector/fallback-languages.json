{
    "ar": {
        "key": "ar",
        "locale": "ar_AE",
        "name": "العربية",
        "localizedName": "Arabic"
    },
    "bg": {
        "key": "bg",
        "locale": "bg_BG",
        "name": "български",
        "localizedName": "Bulgarian"
    },
    "bn": {
        "key": "bn",
        "locale": "bn_BD",
        "name": "বাংলা",
        "localizedName": "Bengali"
    },
    "cs": {
        "key": "cs",
        "locale": "cs_CZ",
        "name": "Čeština",
        "localizedName": "Czech"
    },
    "da": {
        "key": "da",
        "locale": "da_DK",
        "name": "Dansk",
        "localizedName": "Danish"
    },
    "de": {
        "key": "de",
        "locale": "de_DE",
        "name": "Deutsch",
        "localizedName": "German"
    },
    "el": {
        "key": "el",
        "locale": "el_GR",
        "name": "Ελληνικά",
        "localizedName": "Greek"
    },
    "en": {
        "key": "en",
        "locale": "en_US",
        "name": "English",
        "localizedName": "English"
    },
    "es": {
        "key": "es",
        "locale": "es_ES",
        "name": "Español",
        "localizedName": "Spanish"
    },
    "et": {
        "key": "et",
        "locale": "et_EE",
        "name": "Eesti keel",
        "localizedName": "Estonian"
    },
    "fi": {
        "key": "fi",
        "locale": "fi_FI",
        "name": "Suomi",
        "localizedName": "Finnish"
    },
    "fr": {
        "key": "fr",
        "locale": "fr_FR",
        "name": "Français",
        "localizedName": "French"
    },
    "he": {
        "key": "he",
        "locale": "he_IL",
        "name": "עִבְרִית",
        "localizedName": "Hebrew"
    },
    "hi": {
        "key": "hi",
        "locale": "hi_IN",
        "name": "िन्दी",
        "localizedName": "Hindi"
    },
    "hr": {
        "key": "hr",
        "locale": "hr_HR",
        "name": "Hrvatski",
        "localizedName": "Croatian"
    },
    "hu": {
        "key": "hu",
        "locale": "hu_HU",
        "name": "Magyar",
        "localizedName": "Hungarian"
    },
    "is": {
        "key": "is",
        "locale": "is_IS",
        "name": "Íslenska",
        "localizedName": "Icelandic"
    },
    "it": {
        "key": "it",
        "locale": "it_IT",
        "name": "Italiano",
        "localizedName": "Italian"
    },
    "ja": {
        "key": "ja",
        "locale": "ja_JP",
        "name": "日本語",
        "localizedName": "Japanese"
    },
    "ko": {
        "key": "ko",
        "locale": "ko_KR",
        "name": "한국어",
        "localizedName": "Korean"
    },
    "lt": {
        "key": "lt",
        "locale": "lt_LT",
        "name": "Lietuvių",
        "localizedName": "Lithuanian"
    },
    "lv": {
        "key": "lv",
        "locale": "lv_LV",
        "name": "Latviešu",
        "localizedName": "Latvian"
    },
    "mk": {
        "key": "mk",
        "locale": "mk_MK",
        "name": "Македонски",
        "localizedName": "Macedonian"
    },
    "ms": {
        "key": "ms",
        "locale": "ms_MY",
        "name": "Bahasa Melayu",
        "localizedName": "Malay"
    },
    "nl": {
        "key": "nl",
        "locale": "nl_NL",
        "name": "Nederlands",
        "localizedName": "Dutch"
    },
    "no": {
        "key": "no",
        "locale": "no_NO",
        "name": "Norsk",
        "localizedName": "Norwegian"
    },
    "pl": {
        "key": "pl",
        "locale": "pl_PL",
        "name": "Polski",
        "localizedName": "Polish"
    },
    "pt": {
        "key": "pt",
        "locale": "pt_BR",
        "name": "Português",
        "localizedName": "Portuguese"
    },
    "ro": {
        "key": "ro",
        "locale": "ro_RO",
        "name": "Română",
        "localizedName": "Romanian"
    },
    "ru": {
        "key": "ru",
        "locale": "ru_RU",
        "name": "Русский",
        "localizedName": "Russian"
    },
    "sk": {
        "key": "sk",
        "locale": "sk_SK",
        "name": "Slovenský",
        "localizedName": "Slovak"
    },
    "sl": {
        "key": "sl",
        "locale": "sl_SI",
        "name": "Slovenščina",
        "localizedName": "Slovenian"
    },
    "sr": {
        "key": "sr",
        "locale": "sr_RS",
        "name": "Српски",
        "localizedName": "Serbian"
    },
    "sv": {
        "key": "sv",
        "locale": "sv_SE",
        "name": "Svenska",
        "localizedName": "Swedish"
    },
    "th": {
        "key": "th",
        "locale": "th_TH",
        "name": "ภาษาไทย",
        "localizedName": "Thai"
    },
    "tr": {
        "key": "tr",
        "locale": "tr_TR",
        "name": "Türkçe",
        "localizedName": "Turkish"
    },
    "uk": {
        "key": "uk",
        "locale": "uk_UA",
        "name": "Українська",
        "localizedName": "Ukrainian"
    },
    "vi": {
        "key": "vi",
        "locale": "vi_VN",
        "name": "Tiếng Việt",
        "localizedName": "Vietnamese"
    },
    "zh": {
        "key": "zh",
        "locale": "zh_CN",
        "name": "简体中文",
        "localizedName": "Chinese"
    }
}