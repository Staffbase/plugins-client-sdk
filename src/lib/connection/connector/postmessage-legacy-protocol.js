/**
 * Postmessage legacy protocol (3.5)
 */

export default {
  init: 'pluginLoaded', // send this if ready to communicate
  platformInfo: 'platformInfo' // receive this after pluginLoaded was sent
};
