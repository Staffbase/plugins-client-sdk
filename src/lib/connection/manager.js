import genID from './../utils/genId';
import log from 'loglevel';
/**
 * @type {Object.<string, {resolve: function, reject: function, promise: Promise}>}
 */
let promiseMap = {};

/**
 * Create an info object for a new promise in the map.
 *
 * @return {string} id of the promise
 */
function createPromiseObject() {
  const id = genID();

  // When the id is already used, it invokes the function again
  if (id in promiseMap) return createPromiseObject();

  promiseMap[id] = { resolve: null, reject: null, promise: null };
  return id;
}

/**
 * Create a promise and return it's id.
 *
 * The id can be used to operate on the promise using other interface functions.
 *
 * @return {string} id of the promise
 */
export const create = () => {
  const id = createPromiseObject();

  const p = new Promise(function (resolve, reject) {
    promiseMap[id].resolve = resolve;
    promiseMap[id].reject = reject;
  });

  promiseMap[id].promise = p;
  log.debug('promiseManager/create ' + id);
  return id;
};

/**
 * Resolve a promise by id.
 *
 * @param {string} id of the promise
 * @param {any} msg the message which will will be passed to resolve
 */
export const resolve = (id, msg) => {
  log.debug('promiseManager/resolve ' + id);
  if (!(id in promiseMap)) return;

  promiseMap[id].resolve(msg);

  delete promiseMap[id];
};

/**
 * Reject a promise by id.
 *
 * @param {string} id of the promise
 * @param {any} err the error which will will be passed to reject
 */
export const reject = (id, err) => {
  log.debug('promiseManager/reject ' + id);
  if (!(id in promiseMap)) return;

  promiseMap[id].reject(err);
  delete promiseMap[id];
};

/**
 * Get a promise by id.
 *
 * @param {string} id of the promise
 * @return {Promise} the promise identified by id
 * @throws {Error} on unknown id
 */
export const get = (id) => {
  if (!(id in promiseMap)) throw new Error('Tried to get an unknown [' + id + '] promise.');

  return promiseMap[id].promise;
};

/**
 * Resets the state to the default
 */
export const unload = () => {
  promiseMap = {};
};
