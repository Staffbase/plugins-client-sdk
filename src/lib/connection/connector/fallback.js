import { commands as action } from '../commands';
import * as fallbacks from './fallback-handlers';
let log = require('loglevel');

let connection = null;
const fallbackKickIn = 500;

/**
 * Fake connection to the Staffbase App after a period of time.
 *
 * The fallback message handler will signal an established connection
 * after the time specified in fallbackKickIn runs out.
 * @return {Promise<function>} An appropriate send function
 */
export default function connect() {
  if (connection) {
    throw new Error('Connect called twice.');
  }

  connection = new Promise((resolve, reject) => {
    setTimeout(function() {
      log.info('fallback/connect succeeded');
      resolve(sendMessage);
    }, fallbackKickIn);
  });

  return connection;
}

/**
 * Disconnect from the Staffbase App
 *
 * Only usefull for tests.
 */
export function disconnect() {
  connection = null;
}

/**
 * Send a SDK command to the Staffbase App.
 *
 * Maps SDK commands to fallback handlers.
 * @param {String} cmd an SDK command
 * @param {any} payload for the command
 * @return {Promise<any>} which awaits the response of the Staffbase App
 * @throws {Error} on commands not supported by protocol
 */
export async function sendMessage(cmd, ...payload) {
  log.info('fallback/sendMessage ' + cmd);
  log.debug('fallback/sendMessage/payload ' + JSON.stringify(payload));

  switch (cmd) {
    case action.version:
      return fallbacks.getVersion();
    case action.native:
      return fallbacks.isNative();
    case action.mobile:
      return fallbacks.isMobile();
    case action.ios:
      return fallbacks.isIos();
    case action.android:
      return fallbacks.isAndroid();
    case action.openLink:
      return fallbacks.openLink.apply(null, payload);
    case action.nativeUpload:
      return fallbacks.nativeUpload();
    case action.langInfos:
      return fallbacks.langInfos();
    case action.branchDefaultLang:
      return fallbacks.getBranchDefaultLanguage();
    case action.prefContentLang:
      return fallbacks.getPreferredContentLocale.apply(null, payload);
    default:
      // should actualy never ever happen
      throw new Error('Command ' + cmd + ' not supported by driver');
  }
}
