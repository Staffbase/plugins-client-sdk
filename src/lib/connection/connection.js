import postMessageLegacy, {
  disconnect as postMessageLegacyDisconnect
} from './connector/postmessage-legacy.js';
import fallback, { disconnect as fallbackDisconnect } from './connector/fallback.js';
import postMessage, { disconnect as postMessageDisconnect } from './connector/postmessage.js';
import putMessage, { disconnect as putMessageDisconnect } from './connector/putMessage.js';
import { unload as unloadManager } from './manager';
import log from 'loglevel';

let connector;

const connect = async () => {
  const putMessageConnection = putMessage();
  const postMessageConnection = postMessage();
  const postMessagLegacyConnection = postMessageLegacy();
  const fallbackConnection = fallback();

  const realConnectionBucket = [
    putMessageConnection,
    postMessageConnection,
    postMessagLegacyConnection
  ];

  const fallbackConnectionBucket = realConnectionBucket.concat(fallbackConnection);

  // Wait on the real communication and replace the connector with
  Promise.race(realConnectionBucket).then((newConnector) => {
    log.debug('connection/replace connector ' + newConnector.toString());
    connector = newConnector;
  });

  return await Promise.race(fallbackConnectionBucket);
};

export const disconnect = () => {
  postMessageDisconnect();
  putMessageDisconnect();
  postMessageLegacyDisconnect();
  fallbackDisconnect();
  unloadManager();
  connector = null;
};

/**
 * Send a message to the App
 *
 * Will reuse the determined connector
 * to send a message to the Staffbase App
 *
 * @param {any} msg message to send to Staffbase App
 * @param {any} payload that will be attached to the message
 * @return {Promise<any>} result of the request
 */
const sendMessage = async (msg, ...payload) => {
  log.info('connection/sendMessage ' + msg);
  log.debug('connection/sendMessage/payload ' + JSON.stringify(payload));

  if (!connector) {
    connector = connect();
  }

  const sendFn = await connector;
  return sendFn(msg, ...payload);
};

export default sendMessage;
