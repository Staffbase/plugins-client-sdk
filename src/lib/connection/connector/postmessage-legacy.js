import { commands as actions, reversedCommands as reversedActions } from '../commands.js';
import protocol, { invocationMapping } from './postmessage-legacy-protocol.js';
import {
  create as createPromise,
  resolve as resolvePromise,
  get as getPromise
} from '../manager.js';

import { sendMessage as fallBackSendMessage } from './fallback.js';

let log = require('loglevel');

let connection = null;
let connectId = null;
let targetOrigin = '*';

/**
 * @typedef {{ mobile: boolean, version: string|number, native: string }} PlatformInfos
 * @typedef {{ mobile: boolean, version: string|number, native: string, ios: boolean, android: boolean }} StaticValueStore
 */

/**
 * Simple store solution to make the initial data available
 * as static values
 *
 * @param {PlatformInfos} initial the initial data from the frontend
 * @static
 * @return {StaticValueStore}
 */
const dataStore = initial => ({
  mobile: initial.mobile,
  version: initial.version,
  native: !!initial.native,
  ios: initial.native === 'ios',
  android: initial.native === 'android'
});

/**
 * Connect to the Staffbase App.
 *
 * Create a connection to a Staffbase app 3.5
 * @return {Promise<function>} An appropriate send function
 */
const connect = () => {
  if (connection) {
    return connection;
  }

  connectId = createPromise();
  connection = getPromise(connectId).then(function(payload) {
    log.info('pm-legacy/connect succeeded');
    return sendMessage(dataStore(payload));
  });

  window.addEventListener('message', receiveMessage);
  window.parent.postMessage(protocol.init, targetOrigin);
  return connection;
};

export default connect;

/**
 * Disconnect from the Staffbase App
 *
 * Only useful for tests.
 */
export const disconnect = () => {
  window.removeEventListener('message', receiveMessage);
  connection = null;
  connectId = null;
};

/**
 * Handler that receives a message from the Staffbase app
 *
 * Can be attached to window.onPostMessage
 * @param {Object} an onPostMessage event result
 */
const receiveMessage = async ({ data = {} }) => {
  log.info('pm-legacy/receiveMessage ' + data);
  log.debug('pm-legacy/receiveMessage/payload ' + JSON.stringify(data));

  switch (data.state) {
    case protocol.platformInfo:
      log.info('pm-legacy/connect succeeded');
      resolvePromise(connectId, data.info);
      break;
    case protocol.finishUpload:
      log.info('pm-legacy/nativeUpload succeeded');
      resolvePromise(connectId, data.file);
      break;
    default:
      // even thought catch-ignore is a bad style
      // there may be other participants listening
      // to messages in a different format so we
      // silently ignore here
      return;
  }
};

/**
 * Send a SDK command to the Staffbase App.
 *
 * Translates SDK commands into protocol native commands.
 * @param {StaticValueStore} store the store object
 * @param {String} cmd an SDK command
 * @param {array} payload for the command
 * @return {Promise<any>} which awaits the response of the Staffbase App
 * @throws {Error} on commands not supported by protocol
 */
const sendMessage = store => async (cmd, ...payload) => {
  log.info('pm-legacy/sendMessage ' + cmd);
  log.debug('pm-legacy/sendMessage/payload ' + JSON.stringify(payload));

  switch (cmd) {
    case actions.version:
    case actions.native:
    case actions.mobile:
    case actions.ios:
    case actions.android:
      return store[reversedActions[cmd]];
    case actions.nativeUpload:
      return sendInvocationCall(invocationMapping[cmd], payload);
    default:
      return fallBackSendMessage.apply(null, [cmd, ...payload]);
  }
};

/**
 * Create a promise and send an invocation call to the frontend
 *
 * @param {string} process the name of the process to call
 * @param {array} args an array of arguments
 *
 * @return {Promise}
 */
const sendInvocationCall = (process, args) => {
  log.info('pm-legacy/sendInvocationCall ' + process);
  log.debug('pm-legacy/sendInvocationCall/payload ' + JSON.stringify(args));

  connectId = createPromise();
  window.parent.postMessage(protocol.startUpload, targetOrigin);

  return getPromise(connectId);
};
