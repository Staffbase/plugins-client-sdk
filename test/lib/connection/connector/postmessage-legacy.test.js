/* eslint-disable no-global-assign */
/* eslint-env jest, es6 */

import stubPostMessage, { mockPostMessage } from '../../mocks';
import connect from '../../../../src/lib/connection/connector/postmessage-legacy';
import { disconnect } from '../../../../src/lib/connection/connector/postmessage-legacy';
import command from '../../../../src/lib/connection/commands.js';
import { unload as unloadManager } from '../../../../src/lib/connection/manager';

const mockVersion = '3.5-test';
const standardMsg = { state: 'platformInfo', info: {} };

describe('connector/postmessage-legacy', () => {
  describe('connect', () => {
    beforeEach(() => {
      mockPostMessage();
    });

    afterEach(() => {
      disconnect();
      unloadManager();
    });

    it('should be a function', () => {
      expect(connect).toBeFunction();
    });

    it('should yield a Promise', () => {
      const res = connect();
      expect(res instanceof Promise).toBeTrue();
    });

    it('should use postMessage', () => {
      jest.spyOn(window.parent, 'postMessage');
      connect();
      expect(window.parent.postMessage).toHaveBeenCalledTimes(1);
    });

    it('should send a pluginLoaded message', () => {
      jest.spyOn(window.parent, 'postMessage');
      connect();

      expect(window.parent.postMessage).toHaveBeenCalledWith('pluginLoaded', '*');
    });

    it('should resolve on platformInfo message', () => {
      stubPostMessage(standardMsg);
      return expect(connect()).resolves.toBeFunction();
    });

    it('should provide a function', async () => {
      stubPostMessage(standardMsg);
      const fn = await connect();
      return expect(fn).toBeFunction();
    });

    describe('send function', () => {
      describe('accepts sdk commands', () => {
        const info = {
          state: 'platformInfo',
          info: { native: 'ios', mobile: true, version: mockVersion }
        };

        beforeEach(() => {
          stubPostMessage(info);
        });

        it('should reject on unknown commands', async () => {
          const sendFn = await connect();
          try {
            await sendFn('unknown-command');
          } catch (e) {
            expect(e.message).toEqual('Command unknown-command not supported by driver');
          }
        });

        describe('accepts all comands', () => {
          // mock window open
          window.open = function() {};

          const commandData = {
            prefContentLang: ['de_DE', 'en_US']
          };

          for (const cmd in command) {
            if (command.hasOwnProperty(cmd)) {
              it('command.' + cmd, async () => {
                const sendFn = await connect();
                return expect(sendFn(command[cmd], commandData[cmd])).resolves.toMatchSnapshot();
              });
            }
          }
        });
      });

      describe('invoke commands', () => {
        const fileUpload = {
          state: 'finishedImageUploadForPlugin',
          file: true
        };

        beforeEach(() => {
          stubPostMessage(fileUpload);
        });

        it(command.nativeUpload, async () => {
          const sendFn = await connect();
          const nativeUpload = await sendFn(command.nativeUpload);
          expect(nativeUpload).toEqual(true);
        });
      });
    });
  });
});
