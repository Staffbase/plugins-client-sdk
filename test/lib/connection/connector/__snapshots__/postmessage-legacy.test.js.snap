// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`connector/postmessage-legacy connect send function accepts sdk commands accepts all comands command.android 1`] = `false`;

exports[`connector/postmessage-legacy connect send function accepts sdk commands accepts all comands command.branchDefaultLang 1`] = `
Object {
  "key": "en",
  "locale": "en_US",
  "name": "English",
}
`;

exports[`connector/postmessage-legacy connect send function accepts sdk commands accepts all comands command.ios 1`] = `true`;

exports[`connector/postmessage-legacy connect send function accepts sdk commands accepts all comands command.langInfos 1`] = `
Object {
  "branchDefaultLanguage": Object {
    "key": "en",
    "locale": "en_US",
    "name": "English",
  },
  "branchLanguages": Object {
    "ar": Object {
      "key": "ar",
      "locale": "ar_AE",
      "name": "العربية",
    },
    "bg": Object {
      "key": "bg",
      "locale": "bg_BG",
      "name": "български",
    },
    "bn": Object {
      "key": "bn",
      "locale": "bn_BD",
      "name": "বাংলা",
    },
    "cs": Object {
      "key": "cs",
      "locale": "cs_CZ",
      "name": "Čeština",
    },
    "da": Object {
      "key": "da",
      "locale": "da_DK",
      "name": "Dansk",
    },
    "de": Object {
      "key": "de",
      "locale": "de_DE",
      "name": "Deutsch",
    },
    "el": Object {
      "key": "el",
      "locale": "el_GR",
      "name": "Ελληνικά",
    },
    "en": Object {
      "key": "en",
      "locale": "en_US",
      "name": "English",
    },
    "es": Object {
      "key": "es",
      "locale": "es_ES",
      "name": "Español",
    },
    "et": Object {
      "key": "et",
      "locale": "et_EE",
      "name": "Eesti keel",
    },
    "fi": Object {
      "key": "fi",
      "locale": "fi_FI",
      "name": "Suomi",
    },
    "fr": Object {
      "key": "fr",
      "locale": "fr_FR",
      "name": "Français",
    },
    "he": Object {
      "key": "he",
      "locale": "he_IL",
      "name": "עִבְרִית",
    },
    "hi": Object {
      "key": "hi",
      "locale": "hi_IN",
      "name": "िन्दी",
    },
    "hr": Object {
      "key": "hr",
      "locale": "hr_HR",
      "name": "Hrvatski",
    },
    "hu": Object {
      "key": "hu",
      "locale": "hu_HU",
      "name": "Magyar",
    },
    "is": Object {
      "key": "is",
      "locale": "is_IS",
      "name": "Íslenska",
    },
    "it": Object {
      "key": "it",
      "locale": "it_IT",
      "name": "Italiano",
    },
    "ja": Object {
      "key": "ja",
      "locale": "ja_JP",
      "name": "日本語",
    },
    "ko": Object {
      "key": "ko",
      "locale": "ko_KR",
      "name": "한국어",
    },
    "lt": Object {
      "key": "lt",
      "locale": "lt_LT",
      "name": "Lietuvių",
    },
    "lv": Object {
      "key": "lv",
      "locale": "lv_LV",
      "name": "Latviešu",
    },
    "mk": Object {
      "key": "mk",
      "locale": "mk_MK",
      "name": "Македонски",
    },
    "ms": Object {
      "key": "ms",
      "locale": "ms_MY",
      "name": "Bahasa Melayu",
    },
    "nl": Object {
      "key": "nl",
      "locale": "nl_NL",
      "name": "Nederlands",
    },
    "no": Object {
      "key": "no",
      "locale": "no_NO",
      "name": "Norsk",
    },
    "pl": Object {
      "key": "pl",
      "locale": "pl_PL",
      "name": "Polski",
    },
    "pt": Object {
      "key": "pt",
      "locale": "pt_BR",
      "name": "Português",
    },
    "ro": Object {
      "key": "ro",
      "locale": "ro_RO",
      "name": "Română",
    },
    "ru": Object {
      "key": "ru",
      "locale": "ru_RU",
      "name": "Русский",
    },
    "sk": Object {
      "key": "sk",
      "locale": "sk_SK",
      "name": "Slovenský",
    },
    "sl": Object {
      "key": "sl",
      "locale": "sl_SI",
      "name": "Slovenščina",
    },
    "sr": Object {
      "key": "sr",
      "locale": "sr_RS",
      "name": "Српски",
    },
    "sv": Object {
      "key": "sv",
      "locale": "sv_SE",
      "name": "Svenska",
    },
    "th": Object {
      "key": "th",
      "locale": "th_TH",
      "name": "ภาษาไทย",
    },
    "tr": Object {
      "key": "tr",
      "locale": "tr_TR",
      "name": "Türkçe",
    },
    "uk": Object {
      "key": "uk",
      "locale": "uk_UA",
      "name": "Українська",
    },
    "vi": Object {
      "key": "vi",
      "locale": "vi_VN",
      "name": "Tiếng Việt",
    },
    "zh": Object {
      "key": "zh",
      "locale": "zh_CN",
      "name": "简体中文",
    },
  },
  "contentLanguage": Object {
    "key": "en",
    "locale": "en_US",
    "name": "English",
  },
  "contentLanguages": Object {
    "ar": Object {
      "key": "ar",
      "locale": "ar_AE",
      "name": "العربية",
    },
    "bg": Object {
      "key": "bg",
      "locale": "bg_BG",
      "name": "български",
    },
    "bn": Object {
      "key": "bn",
      "locale": "bn_BD",
      "name": "বাংলা",
    },
    "cs": Object {
      "key": "cs",
      "locale": "cs_CZ",
      "name": "Čeština",
    },
    "da": Object {
      "key": "da",
      "locale": "da_DK",
      "name": "Dansk",
    },
    "de": Object {
      "key": "de",
      "locale": "de_DE",
      "name": "Deutsch",
    },
    "el": Object {
      "key": "el",
      "locale": "el_GR",
      "name": "Ελληνικά",
    },
    "en": Object {
      "key": "en",
      "locale": "en_US",
      "name": "English",
    },
    "es": Object {
      "key": "es",
      "locale": "es_ES",
      "name": "Español",
    },
    "et": Object {
      "key": "et",
      "locale": "et_EE",
      "name": "Eesti keel",
    },
    "fi": Object {
      "key": "fi",
      "locale": "fi_FI",
      "name": "Suomi",
    },
    "fr": Object {
      "key": "fr",
      "locale": "fr_FR",
      "name": "Français",
    },
    "he": Object {
      "key": "he",
      "locale": "he_IL",
      "name": "עִבְרִית",
    },
    "hi": Object {
      "key": "hi",
      "locale": "hi_IN",
      "name": "िन्दी",
    },
    "hr": Object {
      "key": "hr",
      "locale": "hr_HR",
      "name": "Hrvatski",
    },
    "hu": Object {
      "key": "hu",
      "locale": "hu_HU",
      "name": "Magyar",
    },
    "is": Object {
      "key": "is",
      "locale": "is_IS",
      "name": "Íslenska",
    },
    "it": Object {
      "key": "it",
      "locale": "it_IT",
      "name": "Italiano",
    },
    "ja": Object {
      "key": "ja",
      "locale": "ja_JP",
      "name": "日本語",
    },
    "ko": Object {
      "key": "ko",
      "locale": "ko_KR",
      "name": "한국어",
    },
    "lt": Object {
      "key": "lt",
      "locale": "lt_LT",
      "name": "Lietuvių",
    },
    "lv": Object {
      "key": "lv",
      "locale": "lv_LV",
      "name": "Latviešu",
    },
    "mk": Object {
      "key": "mk",
      "locale": "mk_MK",
      "name": "Македонски",
    },
    "ms": Object {
      "key": "ms",
      "locale": "ms_MY",
      "name": "Bahasa Melayu",
    },
    "nl": Object {
      "key": "nl",
      "locale": "nl_NL",
      "name": "Nederlands",
    },
    "no": Object {
      "key": "no",
      "locale": "no_NO",
      "name": "Norsk",
    },
    "pl": Object {
      "key": "pl",
      "locale": "pl_PL",
      "name": "Polski",
    },
    "pt": Object {
      "key": "pt",
      "locale": "pt_BR",
      "name": "Português",
    },
    "ro": Object {
      "key": "ro",
      "locale": "ro_RO",
      "name": "Română",
    },
    "ru": Object {
      "key": "ru",
      "locale": "ru_RU",
      "name": "Русский",
    },
    "sk": Object {
      "key": "sk",
      "locale": "sk_SK",
      "name": "Slovenský",
    },
    "sl": Object {
      "key": "sl",
      "locale": "sl_SI",
      "name": "Slovenščina",
    },
    "sr": Object {
      "key": "sr",
      "locale": "sr_RS",
      "name": "Српски",
    },
    "sv": Object {
      "key": "sv",
      "locale": "sv_SE",
      "name": "Svenska",
    },
    "th": Object {
      "key": "th",
      "locale": "th_TH",
      "name": "ภาษาไทย",
    },
    "tr": Object {
      "key": "tr",
      "locale": "tr_TR",
      "name": "Türkçe",
    },
    "uk": Object {
      "key": "uk",
      "locale": "uk_UA",
      "name": "Українська",
    },
    "vi": Object {
      "key": "vi",
      "locale": "vi_VN",
      "name": "Tiếng Việt",
    },
    "zh": Object {
      "key": "zh",
      "locale": "zh_CN",
      "name": "简体中文",
    },
  },
  "deviceLanguage": Object {
    "key": "en",
    "locale": "en_US",
    "name": "English",
  },
}
`;

exports[`connector/postmessage-legacy connect send function accepts sdk commands accepts all comands command.mobile 1`] = `true`;

exports[`connector/postmessage-legacy connect send function accepts sdk commands accepts all comands command.native 1`] = `"ios"`;

exports[`connector/postmessage-legacy connect send function accepts sdk commands accepts all comands command.nativeUpload 1`] = `
Object {
  "mobile": true,
  "native": "ios",
  "version": "3.5-test",
}
`;

exports[`connector/postmessage-legacy connect send function accepts sdk commands accepts all comands command.openLink 1`] = `undefined`;

exports[`connector/postmessage-legacy connect send function accepts sdk commands accepts all comands command.prefContentLang 1`] = `"en_US"`;

exports[`connector/postmessage-legacy connect send function accepts sdk commands accepts all comands command.version 1`] = `"3.5-test"`;
